lib/puppet/provider/foreman_hostgroup/rest_v3.rb

```diff
@@ -0,0 +1,182 @@
+
+    return if success?(r)
+
+    raise Puppet::Error, "Error making POST request to Foreman at #{request_uri(path)}: #{error_message(r)}"
+  end
+
+  def destroy
+    Puppet.debug("Destroying Foreman Hostgroup #{resource[:name]} with parent #{resource[:parent_hostgroup]}")
+    path = "api/v2/hostgroups/#{id}"
+    r = request(:delete, path)
+
+    unless success?(r)
+      error_string = "Error making DELETE request to Foreman at #{request_uri(path)}: #{error_message(r)}"
+      raise Puppet::Error, error_string
+    end
+
+    @hostgroup = nil
+  end
+
+  def flush
+    return if @property_flush.empty?
+
+    Puppet.debug "Calling API to update properties for #{resource[:name]}"
+
+    path = "api/v2/hostgroups/#{id}"
+    r = request(:put, path, {}, { hostgroup: @property_flush }.to_json)
+
+    return if success?(r)
+
+    raise Puppet::Error, "Error making PUT request to Foreman at #{request_uri(path)}: #{error_message(r)}"
+  end
+
+  # Property getters
+  def description
+    hostgroup ? hostgroup['description'] : nil
+  end
+
+  def organizations
+    hostgroup ? hostgroup['organizations'].map { |org| org['title'] } : nil
+  end
+
+  def locations
+    hostgroup ? hostgroup['locations'].map { |org| org['title'] } : nil
+  end
+
+  # Property setters
+  # If one of more properties is being modified then group all of these updates in @property_flush so that we can update them in a single API call in `flush()`
+  def description=(value)
+    @property_flush[:description] = value
+  end
+
+  def organizations=(value)
+    @property_flush[:organization_ids] = value.map { |org| organization_id(org) }
+  end
+
+  def locations=(value)
+    @property_flush[:location_ids] = value.map { |loc| location_id(loc) }
+  end
+
+  private
+
+  def hostgroup
+    @hostgroup ||= begin
+      path = 'api/v2/hostgroups'
+      search_name = resource[:name]
+      search_title = if resource[:parent_hostgroup]
+                       "#{resource[:parent_hostgroup]}/#{resource[:name]}"
+                     else
+                       search_name
+                     end
+      Puppet.debug("Searching for hostgroup with name #{search_name} and title #{search_title}")
+      r = request(:get, path, search: %(title="#{search_title}" and name="#{search_name}"))
+
+      raise Puppet::Error, "Error making GET request to Foreman at #{request_uri(path)}: #{error_message(r)}" unless success?(r)
+
+      results = JSON.parse(r.body)['results']
+      unless results.empty?
+        raise Puppet::Error, "Too many hostgroups found when looking for hostgroup with name #{search_name} and title #{search_title}" if results.size > 1
+
+        get_hostgroup_by_id(results[0]['id'])
+      end
+    end
+  end
+
+  def get_hostgroup_by_id(id)
+    path = "api/v2/hostgroups/#{id}"
+    r = request(:get, path)
+
+    return JSON.parse(r.body) if success?(r)
+
+    raise Puppet::Error, "Error making GET request to Foreman at #{request_uri(path)}: #{error_message(r)}"
```

Comment:
I'm generally not a fan of this pattern. It'll work, but if you at some point need to extend it you'll need to rewrite it. So I prefer to first raise and then "return".
```suggestion
    unless success?(r)
      raise Puppet::Error, "Error making GET request to Foreman at #{request_uri(path)}: #{error_message(r)}"
    end

    JSON.parse(r.body)
```

It's a bit longer, but my reasoning is also that methods should always reach the end unless something abnormally happens.